### DISCUSSION TITLE
WebAssembly Amiga Emulator

### DISCUSSION THREAD

**Discussion by quartexNOR** [dirkwhoffmann/Moira#21] [Ideas]
I read on your website that the codebase is compatible with WASM. Would it be possible to roll a WASM build that can run Amiga software / ADF files et-al?

**Comment by dirkwhoffmann** [comment]
There are already two existing Amiga emulators running in WASM. Both are based on vAmiga which used Moira as it's CPU core. Here are the links: 

- https://github.com/vAmigaWeb/vAmigaWeb
- https://www.neoartcr.com

**Comment by mithrendal** [comment]
@quartexNOR what is your use case ? do you like to embed vAmiga in your website ? vAmigaWeb has a player component. Using this player component you can embed  your ADF on your own web site even without the need of building vAmiga ... simply by placing the player on your web page which then cross loads the emulator files ...  see for reference this how to https://github.com/vAmigaWeb/vAmigaWeb/wiki   (here @rhinoid 
 embedded his own game https://www.proofofconcept.nl/portfolio/hoi/)

of course you can also install the [vAmigaWeb app](https://vamigaweb.github.io/) on your iPhone or iPad homescreen as it complies to the PWA standard i.e. offline installation and more

**Reply by quartexNOR** [reply]
I have spent the past 4 years creating my own development tools (full ide, compiler, rtl et-al) needed to create a cloud variant of Amiga OS. I actually started long before FriendOS and the likes, but I had to implement the devtools which took a bite out of the R&D. The devtool url is: https://www.quartexdeveloper.com if you want to read a bit about it.

Obviously this desktop will run web applications, and is powered by 5 node.js back-end services which provides things like accounts, filesystem over websocket and a unified message system (and tons more). But I had a blast adding DOSBOX to the desktop project, which works great!
And then I figured, why not try to get 68k Amiga stuff running as well.

The original plan was to use SAE, which we modified quite heavily to gain access to the filesystem from the outside. This way, UAE could see our cloud filesystem as a drive, and load/save to it.
SAE works ok, but it's not the fastest, so I was hoping someone would do a WASM build of UAE -- which you have done <3

Then I discovered your system, which seems perfect for the web desktop, allowing me (if possible) to run Amiga software in a web window (window here just being a DIV, which can have any content really).

It really boils down to how easy it is to control the UAE instance from JS, and what it requires. I presume it requires a canvas to render its output only (besides the WASM module etc being loaded and initialized properly).

How much of UAE is exposed in this? Is the pipe used to remote-control it available? Is there an overview of functions i can call to reset, assign binary files (adf et-al) by code, create a snapshot, load rom etc etc -- are there any examples of this?

And great work by the way! It's very smooth!

> @quartexNOR what is your use case ? do you like to embed vAmiga in your website ? vAmigaWeb has a player component. Using this player component you can embed your ADF on your own web site even without the need of building vAmiga ... simply by placing the player on your web page which then cross loads the emulator files ... see for reference this how to https://github.com/vAmigaWeb/vAmigaWeb/wiki (here @rhinoid embedded his own game https://www.proofofconcept.nl/portfolio/hoi/)
> 
> of course you can also install the [vAmigaWeb app](https://vamigaweb.github.io/) on your iPhone or iPad homescreen as it complies to the PWA standard i.e. offline installation and more

**Comment by mithrendal** [comment]
@quartexNOR you are doing cool stuff … I had a look on your dev tool url … looks great ! I will try to give you a short technical summary of the amiga wasm ports and answer your questions

>How much of UAE is exposed in this? Is the pipe used to remote-control it available? Is there an overview of functions i can call to reset, assign binary files (adf et-al) by code, create a snapshot, load rom etc etc -- are there any examples of this?


Moira and vAmiga does not share any code from UAE code base… 

There are currently two open source projects which build vAmiga to wasm 

1. https://github.com/vAmigaWeb/vAmigaWeb 

vAmigaWeb has the following proxy which connects the Javascript/HTML and vAmigaCore/cpp

https://github.com/vAmigaWeb/vAmigaWeb/blob/main/main.cpp

all the methods herein prefixed with `wasm` are exported to javascript

And they are mostly used in https://github.com/vAmigaWeb/vAmigaWeb/blob/main/js/vAmiga_ui.js
https://github.com/vAmigaWeb/vAmigaWeb/blob/main/js/vAmiga_canvas.js
…

vAmigaWeb has a player component which wraps the emulator window in an iFrame which is in turn part of the DOM of the hosting web page… With this player component one can easily  embed vAmiga instances on a webpage … this is easy and clean but due to embedded iFrame the communication of the host page and the vAmigaWeb instance inside the player has to be done via message queue (i.e. browsers postMessage)  because of the browser security police

2. https://github.com/dirkwhoffmann/vAmigaNet

vAmigaNet defines also a proxy which connects the Javascript/HTML and vAmigaCore/cpp

the following file is the cpp part of the proxy…
https://github.com/dirkwhoffmann/vAmigaNet/blob/main/cpp/Proxy.cpp

it is has a svelte counterpart 

https://github.com/dirkwhoffmann/vAmigaNet/blob/main/src/lib/Proxy.svelte

which defines sveltekit stores for the various components of the emulated Amiga.  


vAmigaWeb is a pure javascript / bootstrap PWA (try it at https://vamigaweb.github.io/)

vAmigaNet is a compiled typescript / sveltekit  PWA  (try it at https://vamiganet.github.io/)

**Reply by quartexNOR** [reply]
>> "has to be done via message queue (i.e. browsers postMessage) because of the browser security police"

That is actually perfect, because that is how I host applications right now. When an "app" loads (read: webpage), i wait until the DOM has finished loading, then i post a message to try and establish a link. When the communication is established, i send a secondary messagechannel, which can be used by the app to invoke server-side features. Basicaly a "call" to the server is a JSOn message envelope. The entire system uses the same standard, and the messages are designed to be delegated all the way through the architecture. From the hosted app to the desktop, to the core-server, and further delegated to whatever service has the method that is invoked --- and the reply is routed back the same way :)

Apps that will be honest on the desktop also need a security descriptor, which is an idea i took from android. Its basically a manifest with access rights -- then the user can override and turn things on/off. The user is asked whenever an app try to use the filesystem the first time -- so while cumbersome compared to native code, it kinda works well in a larger setting :)

Thank you for the detailed feedback! I will go over the links and info!

**Comment by mithrendal** [comment]
@quartexNOR sounds really sophisticated… 🤤

Maybe this is also an interesting read for you … it is about communicating with the Amiga …https://github.com/vAmigaWeb/vAmigaWeb/issues/154

from what I understand @PTz0uAH is doing something which goes in a similar direction as what you plan to do… at least when I look at the screenshots which he also posted in a second issue https://github.com/vAmigaWeb/vAmigaWeb/issues/151
